src/main/kotlin/com/github/jengelman/gradle/plugins/shadow/ShadowApplicationPlugin.kt

```diff
@@ -45,18 +45,18 @@ public abstract class ShadowApplicationPlugin : Plugin<Project> {
   }
 
   protected open fun addRunTask() {
-    project.tasks.register(SHADOW_RUN_TASK_NAME, JavaJarExec::class.java) {
+    project.tasks.register(SHADOW_RUN_TASK_NAME, JavaExec::class.java) {
       val install = project.tasks.named(SHADOW_INSTALL_TASK_NAME, Sync::class.java)
+      val jarFile = install.zip(shadowJar) { i, s ->
+        i.destinationDir.resolve("lib/${s.archiveFile.get().asFile.name}")
+      }
+
       it.dependsOn(install)
-      it.mainClass.set("-jar")
+      it.classpath(jarFile)
+      it.mainClass.set(javaApplication.mainClass)
```

Comment:
Set `mainClass` from `javaApplication.mainClass` instead of `-jar` and make sure the `jarFile` is in `classpath`, so that we can get rid of the args appending logic in 

https://github.com/GradleUp/shadow/blob/4850db3f3a9c5755d9d38a43b9337cea0838c4ba/src/main/kotlin/com/github/jengelman/gradle/plugins/shadow/tasks/JavaJarExec.kt#L14-L19